Bonsai.NuGet.Design/IconReader.cs

```diff
@@ -0,0 +1,112 @@
+using System.Drawing.Imaging;
+using System.IO;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Packaging;
+
+namespace Bonsai.NuGet.Design
+{
+    class IconReader
+    {
+        static readonly HttpClient httpClient = new();
+        static readonly Uri PackageDefaultIconUrl = new Uri("https://www.nuget.org/Content/Images/packageDefaultIcon.png");
+        static readonly TimeSpan DefaultIconTimeout = TimeSpan.FromSeconds(10);
+        static readonly Image DefaultIconImage = new Bitmap(32, 32, PixelFormat.Format32bppArgb);
+        readonly ConcurrentDictionary<Uri, Lazy<Task<Image>>> iconCache = new();
+        readonly Task<Image> defaultIcon;
+        readonly Size targetSize;
+
+        public IconReader(Size size)
+        {
+            targetSize = size;
+            defaultIcon = GetAsync(PackageDefaultIconUrl);
+        }
+
+        public void ClearCache()
+        {
+            iconCache.Clear();
+        }
+
+        public Task<Image> GetDefaultIconAsync() => defaultIcon;
+
+        public Task<Image> GetAsync(Uri iconUrl, CancellationToken cancellationToken = default)
+        {
+            if (iconUrl == null) return defaultIcon;
+            if (!iconCache.TryGetValue(iconUrl, out Lazy<Task<Image>> result))
+            {
+                var iconStream = new Lazy<Task<Image>>(() => iconUrl.IsFile
+                    ? GetFileRequestAsync(iconUrl, cancellationToken)
+                    : GetWebRequestAsync(iconUrl, defaultIcon, cancellationToken));
+                result = iconCache.GetOrAdd(iconUrl, iconStream);
+            }
+
+            return result.Value;
+        }
+
+        async Task<Image> GetFileRequestAsync(Uri requestUri, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                if (!string.IsNullOrEmpty(requestUri.Fragment))
+                {
+                    using var packageReader = new PackageArchiveReader(requestUri.AbsolutePath);
+                    using var iconStream = await packageReader.GetStreamAsync(requestUri.Fragment.Substring(1), cancellationToken);
+                    using var image = Image.FromStream(iconStream);
+                    return ResizeImage(image, targetSize);
+                }
+            }
+            catch (IOException) { } // fallback if error reading icon stream
+            catch (ArgumentException) { } // fallback if invalid path or image data
+            catch (UnauthorizedAccessException) { } // fallback if unauthorized access
+
+            return await defaultIcon;
+        }
+
+        async Task<Image> GetWebRequestAsync(Uri requestUri, Task<Image> fallbackImage = null, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                using var requestCancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                requestCancellation.CancelAfter(DefaultIconTimeout);
```

Comment:
Is there a reason this isn't using `HttpClient.Timeout` like `DocumentationHelper`?

Reply:
No reason at all, just that I only learned about it while refactoring `DocumentationHelper` 😅